### ISSUE TITLE
Use the BufferEncoding union defined

### ISSUE THREAD

**Issue by qm3ster** [DefinitelyTyped/DefinitelyTyped#33042]
@SimonSchick told me to raise an additional issue continuing #32926 now that #32878 is merged.

## 1. Use `BufferEncoding` string literal union defined in `globals.d.ts`
First of all, there are at least the following functions:
```ts
Buffer.prototype.write(string: string, encoding?: string) // and overloads
Buffer.prototype.indexOf(string: string, byteOffset?: number, encoding?: string)
Buffer.prototype.lastIndexOf(string: string, byteOffset?: number, encoding?: string)
@deprecated new Buffer(str: string, encoding?: string)
Buffer.from(string: string, encoding?: string)
Buffer.alloc(size: number, fill: string, encoding?: string) // family
Buffer.byteLength(string: string, encoding?: string)
```
As well as on `child_process`, `socket` and `StringEncoder/Decoder` methods that take encoding strings. In all cases the only acceptable values seem to be of that union.

Basically any time `encoding` is mentioned (almost always as `encoding?: string`), it should probably be that union.

## 2. Use overloads instead of just optional parameters
Similarly to what happened in #32926, the `encoding` param is often passed to functions that can take non-strings, such as:
```ts
alloc(size: number, fill?: string | Buffer | number, encoding?: string): Buffer
```
or
```ts
byteLength(string: string | NodeJS.TypedArray | DataView | ArrayBuffer | SharedArrayBuffer, encoding?: string)
```
where encoding is only applicable when preceding parameter is a string

Also, JSDoc on `byteLength` still only talks about strings, so one might get confused that it converts the contents of the other types into the encoding instead of just getting their size.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
